Expose importNode on ShadowRoot and maintain scoped registry association with disconnected nodes

rniwa opened the pull request:
#### 2539a900b51185feffb61785523868aa20ca1482
<pre>
Expose importNode on ShadowRoot and maintain scoped registry association with disconnected nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=283431">https://bugs.webkit.org/show_bug.cgi?id=283431</a>

Reviewed by Chris Dumez.

This PR adds ShadowRoot.prototype.importNode and association of scoped custom element registry with disconnected nodes.
Specifically, when disconnected nodes maintain association with the registry of the last shadow tree to which it belonged.

To maintain the association of an element with a scoped custom element registry, this PR introduces the scoped custom element map.
An element is added to this map whenever it&apos;s synchronously created with a scoped custom element registry or removed from a shadow
tree which uses a scoped custom element registry, and an element is removed from the map when it&apos;s inserted into a shadow tree or
a document tree. An element has ElementStateFlag::UsesScopedCustomElementRegistryMap set when it&apos;s present in this this hash map
so that we can avoid hash map look up in common cases when the element is in a document tree or a shadow tree, or otherwise not
associated with a scoped custom element registry.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-importNode.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-importNode.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-innerHTML.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-innerHTML.tentative.html: Added.

* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::tryToConstructCustomElement): Fixed a bug that we were not resetting the active custom
element registry when the global scope / previous scope wasn&apos;t set.
(WebCore::JSCustomElementInterface::upgradeElement): Ditto.

* Source/WebCore/bindings/js/JSHTMLElementCustom.cpp:
(WebCore::constructJSHTMLElement):

* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::cloneNodeInternal):
* Source/WebCore/dom/Attr.h:
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::cloneNodeInternal):
* Source/WebCore/dom/CDATASection.h:
* Source/WebCore/dom/Comment.cpp:
(WebCore::Comment::cloneNodeInternal):
* Source/WebCore/dom/Comment.h:

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::cloneChildNodes): Now takes TreeScope to clone nodes to.

* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::addToScopedCustomElementRegistryMap): Added.
(WebCore::CustomElementRegistry::removeFromScopedCustomElementRegistryMap): Added.
(WebCore::CustomElementRegistry::scopedCustomElementRegistryMap): Added.

* Source/WebCore/dom/CustomElementRegistry.h:
(WebCore::CustomElementRegistry::isScoped const): Added.
(WebCore::CustomElementRegistry::registryForElement): Added.
(WebCore::CustomElementRegistry::registryForNodeOrTreeScope): Added.

* Source/WebCore/dom/Document.cpp:
(WebCore::createUpgradeCandidateElement): Now takes a TreeScope instead of a Document.
(WebCore::createHTMLElementWithNameValidation):
(WebCore::Document::importNode): Moved to TreeScope.cpp
(WebCore::createFallbackHTMLElement): Now takes a TreeScope instead of a Document.
(WebCore::TreeScope::createElement):
(WebCore::Document::setActiveCustomElementRegistry): Now takes a pointer instead of a reference since we want to able to reset the
value to nullptr when we&apos;re done creating or upgrading an element.
(WebCore::Document::cloneNodeInternal): Now takes a TreeScope instead of a Document.

* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::cloneNodeInternal): Ditto.
(WebCore::DocumentFragment::parseHTML): Now takes CustomElementRegistry as an optional argument.

* Source/WebCore/dom/DocumentFragment.h:
(WebCore::DocumentFragment::parseHTML):

* Source/WebCore/dom/DocumentType.cpp:
(WebCore::DocumentType::cloneNodeInternal):
* Source/WebCore/dom/DocumentType.h:

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::cloneNodeInternal): Now takes a TreeScope instead of a Document.
(WebCore::Element::cloneShadowTreeIfPossible):
(WebCore::Element::cloneElementWithChildren): Ditto.
(WebCore::Element::cloneElementWithoutChildren): Ditto.
(WebCore::Element::cloneElementWithoutAttributesAndChildren): Ditto.
(WebCore::Element::insertedIntoAncestor): Removes the element from the scoped custom element registry map if it&apos;s inserted into
a TreeScope.
(WebCore::Element::removedFromAncestor): Adds the element to the scoped custom element registry map if the registry of the tree
to which this element belonged before the removal is scoped.
(WebCore::Element::replaceChildrenWithMarkup): Calls createFragmentForInnerOuterHTML with an appropriate registry.
(WebCore::Element::setOuterHTML): Ditto.
(WebCore::Element::insertAdjacentHTML): Ditto.

* Source/WebCore/dom/Element.h:

* Source/WebCore/dom/Node.h:
(WebCore::Node::cloneNode):
(WebCore::Node::usesScopedCustomElementRegistryMap const): Added.
(WebCore::Node::setUsesScopedCustomElementRegistryMap): Added.
(WebCore::Node::clearUsesScopedCustomElementRegistryMap): Added.
(WebCore::Node::ElementStateFlag): Added UsesScopedCustomElementRegistryMap as a new enum value.

* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::cloneNodeInternal):
* Source/WebCore/dom/ProcessingInstruction.h:

* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::replaceChildrenWithMarkup): Calls createFragmentForInnerOuterHTML with the custom element registry
associated with this shadow tree.
(WebCore::ShadowRoot::cloneNodeInternal):

* Source/WebCore/dom/ShadowRoot.h:

* Source/WebCore/dom/ShadowRoot.idl:

* Source/WebCore/dom/Text.cpp:
(WebCore::Text::cloneNodeInternal):
* Source/WebCore/dom/Text.h:

* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::importNode): Moved from Document.

* Source/WebCore/dom/TreeScope.h:

* Source/WebCore/editing/markup.cpp:
(WebCore::createFragmentForMarkup): Now optionally takes a custom element registry as an argument.
(WebCore::createFragmentForInnerOuterHTML): Ditto.

* Source/WebCore/editing/markup.h:

* Source/WebCore/html/AttachmentAssociatedElement.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLSourceElement.cpp:
(WebCore::HTMLSourceElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLSourceElement.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::cloneNodeInternal):
* Source/WebCore/html/HTMLTemplateElement.h:

* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::HTMLConstructionSite): Now takes a custom element registry as an argument.
(WebCore::HTMLConstructionSite::createHTMLElementOrFindCustomElementInterface): Add the newly created element to the scoped custom
element map if the registry we used is a scoped custom element registry.

* Source/WebCore/html/parser/HTMLConstructionSite.h:

* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::HTMLDocumentParser): Now takes CustomElementRegistry as an argument.
(WebCore::HTMLDocumentParser::create): Ditto.
(WebCore::HTMLDocumentParser::parseDocumentFragment): Ditto.

* Source/WebCore/html/parser/HTMLDocumentParser.h:
(WebCore::HTMLDocumentParser::parseDocumentFragment): Ditto.

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::HTMLTreeBuilder): Ditto.

* Source/WebCore/html/parser/HTMLTreeBuilder.h:

* Source/WebCore/html/shadow/SliderThumbElement.cpp:
(WebCore::SliderThumbElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/shadow/SliderThumbElement.h:
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::TextTrackCue::create):
(WebCore::TextTrackCue::getCueAsHTML):
(WebCore::TextTrackCue::rebuildDisplayTree):
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCue::createCueRenderingTree):
* Source/WebCore/html/track/WebVTTElement.cpp:
(WebCore::WebVTTElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/track/WebVTTElement.h:
* Source/WebCore/svg/SVGScriptElement.cpp:
(WebCore::SVGScriptElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::cloneDataAndChildren):

Canonical link: <a href="https://commits.webkit.org/287633@main">https://commits.webkit.org/287633@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9e6dc5f68fc7294b75dc28db5a31d42a41021d7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33654 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84758 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31217 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82352 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68304 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7542 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62725 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20537 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83310 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73074 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43033 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50117 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27229 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29678 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27742 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86191 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7461 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5277 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71000 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7636 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68914 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70242 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14236 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13183 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12427 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7425 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12942 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7264 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10784 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9069 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-importNode.tentative.html:
```diff
@@ -0,0 +1,83 @@
+    registry2.define('some-element', SomeElement2);
+    assert_true(shadowRoot2.createElement('some-element') instanceof SomeElement2);
+
+    console.log(shadowRoot2.importNode(someElement));
```
Should we remove this?

rniwa replied:
Oh oops, yeah, this was for debugging.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-importNode.tentative-expected.txt:
```diff
@@ -0,0 +1,6 @@
+CONSOLE MESSAGE: [object HTMLElement]
```
Looks like debugging logic left over?

rniwa replied:
Indeed, removed.

cdumez commented on Source/WebCore/dom/CustomElementRegistry.h:
```diff
@@ -61,8 +64,26 @@ class CustomElementRegistry : public RefCounted<CustomElementRegistry>, public C
+
+    static CustomElementRegistry* registryForNodeOrTreeScope(const Node& node, const TreeScope& treeScope)
+    {
+        if (UNLIKELY(is<Element>(node) && node.usesScopedCustomElementRegistryMap()))
```
Would be nicer to use dynamicDowncast in new code.

rniwa replied:
Sure, we can do that.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1370,20 +1370,23 @@ void Document::childrenChanged(const ChildChange& change)
 
-    Ref element = HTMLMaybeFormAssociatedCustomElement::create(name, document);
+    Ref element = HTMLMaybeFormAssociatedCustomElement::create(name, treeScope.documentScope());
+    auto* registry = treeScope.customElementRegistry();
```
Would be nice to use a smart pointer here.

rniwa replied:
Sure, we can use RefPtr.

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3002,6 +3004,12 @@ void Element::removedFromAncestor(RemovalType removalType, ContainerNode& oldPar
                 updateNameForDocument(*oldHTMLDocument, nameValue, nullAtom());
         }
+        if (oldParentOfRemovedTree.isInShadowTree()) {
+            if (auto* registry = oldTreeScope.customElementRegistry()) {
```
Smart pointer?

rniwa replied:
Sure, deployed.

cdumez reviewed (approved):
LGTM % nits.

rniwa committed:
Expose importNode on ShadowRoot and maintain scoped registry association with disconnected nodes
https://bugs.webkit.org/show_bug.cgi?id=283431

Reviewed by Chris Dumez.

This PR adds ShadowRoot.prototype.importNode and association of scoped custom element registry with disconnected nodes.
Specifically, when disconnected nodes maintain association with the registry of the last shadow tree to which it belonged.

To maintain the association of an element with a scoped custom element registry, this PR introduces the scoped custom element map.
An element is added to this map whenever it's synchronously created with a scoped custom element registry or removed from a shadow
tree which uses a scoped custom element registry, and an element is removed from the map when it's inserted into a shadow tree or
a document tree. An element has ElementStateFlag::UsesScopedCustomElementRegistryMap set when it's present in this this hash map
so that we can avoid hash map look up in common cases when the element is in a document tree or a shadow tree, or otherwise not
associated with a scoped custom element registry.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-importNode.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-importNode.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-innerHTML.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/scoped-registry/ShadowRoot-innerHTML.tentative.html: Added.

* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::tryToConstructCustomElement): Fixed a bug that we were not resetting the active custom
element registry when the global scope / previous scope wasn't set.
(WebCore::JSCustomElementInterface::upgradeElement): Ditto.

* Source/WebCore/bindings/js/JSHTMLElementCustom.cpp:
(WebCore::constructJSHTMLElement):

* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::cloneNodeInternal):
* Source/WebCore/dom/Attr.h:
* Source/WebCore/dom/CDATASection.cpp:
(WebCore::CDATASection::cloneNodeInternal):
* Source/WebCore/dom/CDATASection.h:
* Source/WebCore/dom/Comment.cpp:
(WebCore::Comment::cloneNodeInternal):
* Source/WebCore/dom/Comment.h:

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::cloneChildNodes): Now takes TreeScope to clone nodes to.

* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::addToScopedCustomElementRegistryMap): Added.
(WebCore::CustomElementRegistry::removeFromScopedCustomElementRegistryMap): Added.
(WebCore::CustomElementRegistry::scopedCustomElementRegistryMap): Added.

* Source/WebCore/dom/CustomElementRegistry.h:
(WebCore::CustomElementRegistry::isScoped const): Added.
(WebCore::CustomElementRegistry::registryForElement): Added.
(WebCore::CustomElementRegistry::registryForNodeOrTreeScope): Added.

* Source/WebCore/dom/Document.cpp:
(WebCore::createUpgradeCandidateElement): Now takes a TreeScope instead of a Document.
(WebCore::createHTMLElementWithNameValidation):
(WebCore::Document::importNode): Moved to TreeScope.cpp
(WebCore::createFallbackHTMLElement): Now takes a TreeScope instead of a Document.
(WebCore::TreeScope::createElement):
(WebCore::Document::setActiveCustomElementRegistry): Now takes a pointer instead of a reference since we want to able to reset the
value to nullptr when we're done creating or upgrading an element.
(WebCore::Document::cloneNodeInternal): Now takes a TreeScope instead of a Document.

* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::cloneNodeInternal): Ditto.
(WebCore::DocumentFragment::parseHTML): Now takes CustomElementRegistry as an optional argument.

* Source/WebCore/dom/DocumentFragment.h:
(WebCore::DocumentFragment::parseHTML):

* Source/WebCore/dom/DocumentType.cpp:
(WebCore::DocumentType::cloneNodeInternal):
* Source/WebCore/dom/DocumentType.h:

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::cloneNodeInternal): Now takes a TreeScope instead of a Document.
(WebCore::Element::cloneShadowTreeIfPossible):
(WebCore::Element::cloneElementWithChildren): Ditto.
(WebCore::Element::cloneElementWithoutChildren): Ditto.
(WebCore::Element::cloneElementWithoutAttributesAndChildren): Ditto.
(WebCore::Element::insertedIntoAncestor): Removes the element from the scoped custom element registry map if it's inserted into
a TreeScope.
(WebCore::Element::removedFromAncestor): Adds the element to the scoped custom element registry map if the registry of the tree
to which this element belonged before the removal is scoped.
(WebCore::Element::replaceChildrenWithMarkup): Calls createFragmentForInnerOuterHTML with an appropriate registry.
(WebCore::Element::setOuterHTML): Ditto.
(WebCore::Element::insertAdjacentHTML): Ditto.

* Source/WebCore/dom/Element.h:

* Source/WebCore/dom/Node.h:
(WebCore::Node::cloneNode):
(WebCore::Node::usesScopedCustomElementRegistryMap const): Added.
(WebCore::Node::setUsesScopedCustomElementRegistryMap): Added.
(WebCore::Node::clearUsesScopedCustomElementRegistryMap): Added.
(WebCore::Node::ElementStateFlag): Added UsesScopedCustomElementRegistryMap as a new enum value.

* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::cloneNodeInternal):
* Source/WebCore/dom/ProcessingInstruction.h:

* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::replaceChildrenWithMarkup): Calls createFragmentForInnerOuterHTML with the custom element registry
associated with this shadow tree.
(WebCore::ShadowRoot::cloneNodeInternal):

* Source/WebCore/dom/ShadowRoot.h:

* Source/WebCore/dom/ShadowRoot.idl:

* Source/WebCore/dom/Text.cpp:
(WebCore::Text::cloneNodeInternal):
* Source/WebCore/dom/Text.h:

* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::importNode): Moved from Document.

* Source/WebCore/dom/TreeScope.h:

* Source/WebCore/editing/markup.cpp:
(WebCore::createFragmentForMarkup): Now optionally takes a custom element registry as an argument.
(WebCore::createFragmentForInnerOuterHTML): Ditto.

* Source/WebCore/editing/markup.h:

* Source/WebCore/html/AttachmentAssociatedElement.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLSourceElement.cpp:
(WebCore::HTMLSourceElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLSourceElement.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::cloneNodeInternal):
* Source/WebCore/html/HTMLTemplateElement.h:

* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::HTMLConstructionSite): Now takes a custom element registry as an argument.
(WebCore::HTMLConstructionSite::createHTMLElementOrFindCustomElementInterface): Add the newly created element to the scoped custom
element map if the registry we used is a scoped custom element registry.

* Source/WebCore/html/parser/HTMLConstructionSite.h:

* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::HTMLDocumentParser): Now takes CustomElementRegistry as an argument.
(WebCore::HTMLDocumentParser::create): Ditto.
(WebCore::HTMLDocumentParser::parseDocumentFragment): Ditto.

* Source/WebCore/html/parser/HTMLDocumentParser.h:
(WebCore::HTMLDocumentParser::parseDocumentFragment): Ditto.

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::HTMLTreeBuilder): Ditto.

* Source/WebCore/html/parser/HTMLTreeBuilder.h:

* Source/WebCore/html/shadow/SliderThumbElement.cpp:
(WebCore::SliderThumbElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/shadow/SliderThumbElement.h:
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::TextTrackCue::create):
(WebCore::TextTrackCue::getCueAsHTML):
(WebCore::TextTrackCue::rebuildDisplayTree):
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCue::createCueRenderingTree):
* Source/WebCore/html/track/WebVTTElement.cpp:
(WebCore::WebVTTElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/track/WebVTTElement.h:
* Source/WebCore/svg/SVGScriptElement.cpp:
(WebCore::SVGScriptElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::cloneDataAndChildren):

Canonical link: https://commits.webkit.org/287633@main
